docs/reference/filesystem.rst

```diff
@@ -0,0 +1,272 @@
+
+The following table shows which filesystems are included in the firmware by
+default for given port/board combinations, however they can be optionally
+enabled in a custom firmware build.
+
+====================  =====  ===========  ===========
+Board                 FAT    littlefs v1  littlefs v2
+====================  =====  ===========  ===========
+pyboard 1.0, 1.1, D   Yes    No           Yes
+Other STM32           Yes    No           No
+ESP8266               Yes    No           No
+ESP32                 Yes    No           Yes
+====================  =====  ===========  ===========
+
+FAT
+~~~
+
+The main advantage of the FAT filesystem is that it can be accessed over USB MSC
+on supported boards (e.g. STM32) without any additional drivers required on the
+host PC.
+
+However, FAT is not tolerant to power failure during writes and this can lead to
+filesystem corruption. For applications that do not require USB MSC, it is
+recommended to use littlefs instead.
+
+To format the entire flash using FAT::
+
+    # ESP8266 and ESP32
+    import os
+    os.umount('/')
+    os.VfsFat.mkfs(bdev)
+    os.mount(bdev, '/')
+
+    # STM32
+    import os, pyb
+    os.umount('/flash')
+    os.VfsFat.mkfs(pyb.Flash(start=0))
+    os.mount(pyb.Flash(start=0), '/flash')
+    os.chdir('/flash')
+
+Littlefs
+~~~~~~~~
+
+Littlefs_ is a filesystem designed for flash-based devices, and is much more
+resistant to filesystem corruption.
+
+Note: It can be still be accessed over USB MSC using the `littlefs FUSE
+driver`_. Note that you must use the ``-b=4096`` option to override the block
+size.
+
+.. _littlefs FUSE driver: https://github.com/ARMmbed/littlefs-fuse/tree/master/littlefs
+
+.. _Littlefs: https://github.com/ARMmbed/littlefs
+
+To format the entire flash using littlefs v2::
+
+    # ESP8266 and ESP32
+    import os
+    os.umount('/')
+    os.VfsLfs2.mkfs(bdev)
+    os.mount(bdev, '/')
+
+    # STM32
+    import os, pyb
+    os.umount('/flash')
+    os.VfsLfs2.mkfs(pyb.Flash(start=0))
+    os.mount(pyb.Flash(start=0), '/flash')
+    os.chdir('/flash')
+
+Hybrid (STM32)
+~~~~~~~~~~~~~~
+
+By using the ``start`` and ``len`` kwargs to :class:`pyb.Flash`, you can create
+block devices spanning a subset of the flash device.
+
+For example, to configure the first 256kiB as FAT (and available over USB MSC),
+and the remainder as littlefs::
+
+    import os, pyb
+    os.umount('/flash')
+    p1 = pyb.Flash(start=0, len=256*1024)
+    p2 = pyb.Flash(start=256*1024)
+    os.VfsFat.mkfs(p1)
+    os.VfsLfs2.mkfs(p2)
+    os.mount(p1, '/flash')
+    os.mount(p2, '/data')
+    os.chdir('/flash')
+
+This might be useful to make your Python files, configuration and other
+rarely-modified content available over USB MSC, but allowing for frequently
+changing application data to reside on littlefs with better resiliency to power
+failure, etc.
+
+The partition at offset ``0`` will be mounted automatically, but you can add::
```

Comment:
"... will be mounted automatically (and the filesystem type automatically detected), ..."